##**RUSH Sportswear Company - Sales Data Analysis**

####by Jeff Hansen


####Importing libraries and getting the data

In [8]:
# import necessary libraries
import pandas as pd

In [9]:
# read data into pandas data frames
df_sales = pd.read_csv('TABLE_SALES_885.csv')

df_retailer = pd.read_csv('TABLE_RETAILER_885.csv')

df_products = pd.read_csv('TABLE_PRODUCTS_885.csv', sep='|')

####**Inspecting the data**
#### 3 dataframes for the 3 CSV files

In [10]:
# inspect df_sales
df_sales.head()

,ORDER_ID,RETAILER_ID,INVOICE_DATE,MONTH,DAY,YEAR,PRODUCT_ID,PRICE_PER_UNIT,UNITS_SOLD,OPERATING_MARGIN,SALES_METHOD
0,1,A00MOHCO,1/1/2020,1,1,2020,20,50.0,1200,0.5,In-store
1,7,A00MOHCO,1/7/2020,1,7,2020,20,50.0,1250,0.5,In-store
2,13,A00MOHCO,1/25/2020,1,25,2020,20,50.0,1220,0.5,Outlet
3,19,A00MOHCO,1/31/2020,1,31,2020,20,50.0,1200,0.5,Outlet
4,25,A00MOHCO,2/6/2020,2,6,2020,20,60.0,1220,0.5,Outlet


In [11]:
# inspecting df_sales categorical fields, looking for misspellings RESULT: Ootlet is incorrect
for column in ['SALES_METHOD']:
    print(f"\n{column}")
    print(df_sales[column].value_counts())


SALES_METHOD
SALES_METHOD
Online      4889
Outlet      2999
In-store    1740
Ootlet        20
Name: count, dtype: int64


In [12]:
# exploring df_sales
# note the two null values in PRICE_PER_UNIT, the object data type for INVOICE_DATE
# and the object data type for UNITS_SOLD
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9648 entries, 0 to 9647
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ORDER_ID          9648 non-null   int64  
 1   RETAILER_ID       9648 non-null   object 
 2   INVOICE_DATE      9648 non-null   object 
 3   MONTH             9648 non-null   int64  
 4   DAY               9648 non-null   int64  
 5   YEAR              9648 non-null   int64  
 6   PRODUCT_ID        9648 non-null   int64  
 7   PRICE_PER_UNIT    9646 non-null   float64
 8   UNITS_SOLD        9648 non-null   object 
 9   OPERATING_MARGIN  9648 non-null   float64
 10  SALES_METHOD      9648 non-null   object 
dtypes: float64(2), int64(5), object(4)
memory usage: 829.3+ KB


In [13]:
# in spect df_sales
df_sales.shape

(9648, 11)

In [14]:
# how many times does a distinct value appear in UNITS_SOLD? RESULT:
df_sales['UNITS_SOLD'].value_counts(dropna=False)

,count
UNITS_SOLD,
225,207
150,202
175,191
300,189
200,174
...,...
16,1
286,1
347,1


In [15]:
# if I treat UNITS_SOLD as numeric, how many would fail? RESULT: 2
pd.to_numeric(df_sales['UNITS_SOLD'], errors='coerce').isna().sum()

np.int64(2)

In [16]:
# which rows would fail? i.e. this will explain why UNITS_SOLD got stored as an object data type
df_sales[
    pd.to_numeric(df_sales['UNITS_SOLD'], errors='coerce').isna()
]

,ORDER_ID,RETAILER_ID,INVOICE_DATE,MONTH,DAY,YEAR,PRODUCT_ID,PRICE_PER_UNIT,UNITS_SOLD,OPERATING_MARGIN,SALES_METHOD
1012,6064,S00SALBI,5/27/2021,5,27,2021,20,51.0,***,0.45,Online
1439,8626,W00MIODE,12/10/2021,12,10,2021,20,29.0,***,0.46,Outlet


In [17]:
# looking at PRIE_PER_UNIT now and verifying there are 2 nulls
df_sales['PRICE_PER_UNIT'].isna().sum()

np.int64(2)

In [18]:
# which rows have a null PRICE_PER_UNIT and what does it look like? RESULT: NaN
df_sales[df_sales['PRICE_PER_UNIT'].isna()]

,ORDER_ID,RETAILER_ID,INVOICE_DATE,MONTH,DAY,YEAR,PRODUCT_ID,PRICE_PER_UNIT,UNITS_SOLD,OPERATING_MARGIN,SALES_METHOD
98,591,A00NVEBU,4/2/2020,4,2,2020,20,NaN,525,0.35,In-store
99,597,A00NVEBU,4/8/2020,4,8,2020,20,NaN,525,0.50,In-store


In [19]:
# checking for duplicates. RESULT: there are NONE
df_sales.duplicated().sum()

np.int64(0)

In [20]:
# inspecting INVOICE_DATE. RESULT: nothing looks weird
df_sales['INVOICE_DATE'].head(10)

,INVOICE_DATE
0,1/1/2020
1,1/7/2020
2,1/25/2020
3,1/31/2020
4,2/6/2020
5,3/4/2020
6,3/10/2020
7,3/16/2020
8,4/19/2020
9,4/27/2020


In [21]:
# any missing INVOICE_DATES? RESULT: There are NONE.
df_sales['INVOICE_DATE'].isna().sum()

np.int64(0)

In [22]:
# moving on to df_retailer and inspecting. RESULT: Looks good
df_retailer.head()

,RETAILER_ID,RETAILER,REGION,STATE,CITY
0,A00MOHCO,Amazon,Midwest,Ohio,Columbus
1,A00NMAPO,Amazon,Northeast,Maine,Portland
2,A00NMABO,Amazon,Northeast,Massachusetts,Boston
3,A00NNEMA,Amazon,Northeast,New Hampshire,Manchester
4,A00NVEBU,Amazon,Northeast,Vermont,Burlington


In [23]:
# inspecting df_retailer. RESULT: 5 columns, all objects, no nulls
df_retailer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110 entries, 0 to 109
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   RETAILER_ID  110 non-null    object
 1   RETAILER     110 non-null    object
 2   REGION       110 non-null    object
 3   STATE        110 non-null    object
 4   CITY         110 non-null    object
dtypes: object(5)
memory usage: 4.4+ KB


In [24]:
# inspecting df_retailer. RESULT: 110 rows 5 columns
df_retailer.shape

(110, 5)

In [25]:
# inspecting df_retailer categorical columns for weirdness
# RESULT: Not spotting anything weird
for column in ['RETAILER', 'REGION', 'STATE', 'CITY']:
    print(f"\n{column}")
    print(df_retailer[column].value_counts())


RETAILER
RETAILER
Foot Locker      33
Sports Direct    26
West Gear        24
Kohl's           12
Amazon            9
Walmart           6
Name: count, dtype: int64

REGION
REGION
Northeast    28
Midwest      24
West         24
South        18
Southeast    16
Name: count, dtype: int64

STATE
STATE
Florida           6
New York          5
Texas             5
California        4
Idaho             3
Mississippi       3
Maryland          3
Hawaii            3
Alaska            2
Illinois          2
Massachusetts     2
New Hampshire     2
Vermont           2
Alabama           2
Kentucky          2
North Carolina    2
Ohio              2
Maine             2
South Dakota      2
North Dakota      2
Nebraska          2
Missouri          2
Minnesota         2
Michigan          2
Kansas            2
Iowa              2
Louisiana         2
West Virginia     2
Rhode Island      2
Tennessee         2
Georgia           2
Pennsylvania      2
Delaware          2
Connecticut       2
Arizona           2
V

In [26]:
# looking for duplicates in df_retailer. RESULT: as expected, there are no duplicates
df_retailer.duplicated().sum()

np.int64(0)

In [27]:
# any RETAILER_IDs used more than once? RESULT: Yes, 4
df_retailer['RETAILER_ID'].duplicated().sum()

np.int64(4)

In [28]:
# inspecting the rows where RETAILER_ID is repeated
# RETAILER_ID contains 4 duplicated values. Some duplicated IDs correspond
# to different retailer names or locations, so the key is not fully unique
# and must be reviewed before merging with sales data.
df_retailer[
    df_retailer['RETAILER_ID'].duplicated(keep=False)
].sort_values('RETAILER_ID')

,RETAILER_ID,RETAILER,REGION,STATE,CITY
63,S00NNENE,Sports Direct,Northeast,New Jersey,Newark
64,S00NNENE,Sports Direct,Northeast,New York,New York
81,W00SARLI,Walmart,South,Arkansas,Little Rock
97,W00SARLI,West Gear,South,Arkansas,Little Rock
84,W00SFLOR,Walmart,Southeast,Florida,Orlando
102,W00SFLOR,West Gear,Southeast,Florida,Orlando
83,W00STEHO,Walmart,South,Texas,Houston
100,W00STEHO,West Gear,South,Texas,Houston


In [29]:
# how many records for each retailer? RESULT: Foot Locker has 33,
# Sports Direct 26, West Gear 24, Kohls 12, Amazon 9, Walmart 6
df_retailer['RETAILER'].value_counts()

,count
RETAILER,
Foot Locker,33
Sports Direct,26
West Gear,24
Kohl's,12
Amazon,9
Walmart,6


In [30]:
# now inspecting the products table. RESULT: loaded correctly
df_products.head()

,PRODUCT_ID,PRODUCT_NAME
0,20,Men's Street Footwear
1,30,Men's Athletic Footwear
2,120,Women's Street Footwear
3,130,Women's Athletic Footwear
4,40,Men's Apparel


In [31]:
# inspecting products table. RESULT: 6 rows, 2 columns, no duplicates,
# no missing values, no weird data types
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PRODUCT_ID    6 non-null      int64 
 1   PRODUCT_NAME  6 non-null      object
dtypes: int64(1), object(1)
memory usage: 228.0+ bytes


In [32]:
# checking for dulicates in df_products. RESULT = 0
df_products.duplicated().sum()

np.int64(0)

In [33]:
# is PRODUCT_ID unique? RESULT: yes
df_products['PRODUCT_ID'].duplicated().sum()

np.int64(0)

###**Data cleaning**

####Cleaning df_sales

In [34]:
df_sales[
    pd.to_numeric(df_sales['UNITS_SOLD'], errors='coerce').isna()
]

,ORDER_ID,RETAILER_ID,INVOICE_DATE,MONTH,DAY,YEAR,PRODUCT_ID,PRICE_PER_UNIT,UNITS_SOLD,OPERATING_MARGIN,SALES_METHOD
1012,6064,S00SALBI,5/27/2021,5,27,2021,20,51.0,***,0.45,Online
1439,8626,W00MIODE,12/10/2021,12,10,2021,20,29.0,***,0.46,Outlet


In [35]:
# looking at UNITS_SOLD for PRODUCT_ID = 20 in year 2021
# RESULT median = 232, mean = 355, high variability --> use median
df_sales[
    (df_sales['PRODUCT_ID'] == 20) &
    (df_sales['YEAR'] == 2021) &
    (df_sales['UNITS_SOLD'] != '***')
]['UNITS_SOLD'].astype(int).describe()

,UNITS_SOLD
count,1389.000000
mean,355.473722
std,239.417642
min,84.000000
25%,163.000000
50%,232.000000
75%,525.000000
max,1150.000000


In [36]:
# replacing the ... with median = 232
df_sales.loc[
    df_sales['UNITS_SOLD'] == '***',
    'UNITS_SOLD'
] = 232

In [37]:
# now we can convert UNITS_SOLD to numeric
df_sales['UNITS_SOLD'] = pd.to_numeric(df_sales['UNITS_SOLD'])

In [38]:
# verify the result (UNITS_SOLD is not an int64)
df_sales['UNITS_SOLD'].dtype

dtype('int64')

In [39]:
# verify there are no more invalid dates. RESULT = 0
pd.to_numeric(df_sales['UNITS_SOLD'], errors='coerce').isna().sum()

np.int64(0)

Two missing PRICE_PER_UNIT values were identified for Product 20 in 2020. After reviewing nearby comparable transactions for the same retailer and sales method, both missing values were replaced with $50.

In [40]:
# clean the missing PRICE_PER_UNIT
df_sales[df_sales['PRICE_PER_UNIT'].isna()]

,ORDER_ID,RETAILER_ID,INVOICE_DATE,MONTH,DAY,YEAR,PRODUCT_ID,PRICE_PER_UNIT,UNITS_SOLD,OPERATING_MARGIN,SALES_METHOD
98,591,A00NVEBU,4/2/2020,4,2,2020,20,NaN,525,0.35,In-store
99,597,A00NVEBU,4/8/2020,4,8,2020,20,NaN,525,0.50,In-store


In [41]:
# exploring options for the missing PRICE_PER_UNIT observations
df_sales[df_sales['PRICE_PER_UNIT'].isna()][
    ['PRODUCT_ID', 'YEAR', 'RETAILER_ID', 'UNITS_SOLD', 'SALES_METHOD']
]

,PRODUCT_ID,YEAR,RETAILER_ID,UNITS_SOLD,SALES_METHOD
98,20,2020,A00NVEBU,525,In-store
99,20,2020,A00NVEBU,525,In-store


In [42]:
# exploring options for the missing PRICE_PER_UNIT observations
df_sales[
    (df_sales['PRODUCT_ID'] == 20) &
    (df_sales['YEAR'] == 2020) &
    (df_sales['RETAILER_ID'] == 'A00NVEBU') &
    (df_sales['SALES_METHOD'] == 'In-store') &
    (df_sales['PRICE_PER_UNIT'].notna())
][['PRICE_PER_UNIT', 'UNITS_SOLD']].describe()

,PRICE_PER_UNIT,UNITS_SOLD
count,21.000000,21.000000
mean,55.952381,602.380952
std,12.905333,97.116155
min,30.000000,425.000000
25%,50.000000,550.000000
50%,60.000000,575.000000
75%,65.000000,700.000000
max,75.000000,775.000000


In [43]:
# exploring options for the missing PRICE_PER_UNIT observations
df_sales[
    (df_sales['PRODUCT_ID'] == 20) &
    (df_sales['YEAR'] == 2020) &
    (df_sales['RETAILER_ID'] == 'A00NVEBU') &
    (df_sales['SALES_METHOD'] == 'In-store') &
    (df_sales['UNITS_SOLD'] == 525) &
    (df_sales['PRICE_PER_UNIT'].notna())
][['INVOICE_DATE', 'PRICE_PER_UNIT', 'UNITS_SOLD']]

,INVOICE_DATE,PRICE_PER_UNIT,UNITS_SOLD
96,3/21/2020,30.0,525
104,5/8/2020,55.0,525


In [44]:
# convert INVOICE_DATE to datetime
df_sales['INVOICE_DATE'] = pd.to_datetime(df_sales['INVOICE_DATE'])

In [45]:
# confirming result
df_sales['INVOICE_DATE'].dtype

dtype('<M8[ns]')

In [46]:
# exploring options for imputing missing PRICE_PER_UNIT
df_sales[
    (df_sales['PRODUCT_ID'] == 20) &
    (df_sales['YEAR'] == 2020) &
    (df_sales['RETAILER_ID'] == 'A00NVEBU') &
    (df_sales['SALES_METHOD'] == 'In-store')
][
    ['INVOICE_DATE', 'PRICE_PER_UNIT', 'UNITS_SOLD']
].sort_values('INVOICE_DATE')

,INVOICE_DATE,PRICE_PER_UNIT,UNITS_SOLD
90,2020-02-13,65.0,700
91,2020-02-19,60.0,600
92,2020-02-25,60.0,550
93,2020-03-03,60.0,575
94,2020-03-09,60.0,675
95,2020-03-15,30.0,425
96,2020-03-21,30.0,525
97,2020-03-27,40.0,550
98,2020-04-02,NaN,525
99,2020-04-08,NaN,525


In [47]:
# imputing 50 (midpoint between 40 and 60)
df_sales['PRICE_PER_UNIT'] = df_sales['PRICE_PER_UNIT'].fillna(50)

In [48]:
# verifying result
df_sales['PRICE_PER_UNIT'].isna().sum()

np.int64(0)

In [49]:
# fixing SALES_METHOD = Ootlet
df_sales['SALES_METHOD'] = df_sales['SALES_METHOD'].replace('Ootlet', 'Outlet')

In [50]:
# verifying the result
df_sales['SALES_METHOD'].value_counts()

,count
SALES_METHOD,
Online,4889
Outlet,3019
In-store,1740


In [51]:
# inspecting df_sales again
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9648 entries, 0 to 9647
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ORDER_ID          9648 non-null   int64         
 1   RETAILER_ID       9648 non-null   object        
 2   INVOICE_DATE      9648 non-null   datetime64[ns]
 3   MONTH             9648 non-null   int64         
 4   DAY               9648 non-null   int64         
 5   YEAR              9648 non-null   int64         
 6   PRODUCT_ID        9648 non-null   int64         
 7   PRICE_PER_UNIT    9648 non-null   float64       
 8   UNITS_SOLD        9648 non-null   int64         
 9   OPERATING_MARGIN  9648 non-null   float64       
 10  SALES_METHOD      9648 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(6), object(2)
memory usage: 829.3+ KB


In [52]:
df_sales.isna().sum()

,0
ORDER_ID,0
RETAILER_ID,0
INVOICE_DATE,0
MONTH,0
DAY,0
YEAR,0
PRODUCT_ID,0
PRICE_PER_UNIT,0
UNITS_SOLD,0
OPERATING_MARGIN,0


####Cleaning df_retailer

In [53]:
# cleaning df_retailer
duplicate_retailer_ids = df_retailer[
    df_retailer['RETAILER_ID'].duplicated(keep=False)
]['RETAILER_ID'].unique()

df_sales[
    df_sales['RETAILER_ID'].isin(duplicate_retailer_ids)
]['RETAILER_ID'].value_counts()

,count
RETAILER_ID,
W00SARLI,216
W00SFLOR,189
W00STEHO,162
S00NNENE,56


In [54]:
# dealing with ambiguity re: retailer
df_sales[
    df_sales['RETAILER_ID'] == 'W00SARLI'
].head(20)

,ORDER_ID,RETAILER_ID,INVOICE_DATE,MONTH,DAY,YEAR,PRODUCT_ID,PRICE_PER_UNIT,UNITS_SOLD,OPERATING_MARGIN,SALES_METHOD
1164,6976,W00SARLI,2021-01-03,1,3,2021,20,28.0,126,0.45,Online
1165,6982,W00SARLI,2021-02-03,2,3,2021,20,29.0,140,0.41,Online
1166,6988,W00SARLI,2021-03-02,3,2,2021,20,19.0,141,0.41,Online
1167,6994,W00SARLI,2021-04-03,4,3,2021,20,18.0,122,0.41,Online
1168,7000,W00SARLI,2021-05-04,5,4,2021,20,58.0,134,0.41,Online
1169,7006,W00SARLI,2021-06-03,6,3,2021,20,32.0,160,0.44,Online
1170,7012,W00SARLI,2021-07-02,7,2,2021,20,45.0,149,0.42,Online
1171,7018,W00SARLI,2021-08-03,8,3,2021,20,46.0,158,0.42,Online
1172,7024,W00SARLI,2021-09-04,9,4,2021,20,33.0,143,0.41,Online
1173,7030,W00SARLI,2021-10-03,10,3,2021,20,43.0,113,0.42,Online


In [55]:
df_retailer[
    (df_retailer['STATE'] == 'Arkansas') &
    (df_retailer['CITY'] == 'Little Rock')
]

,RETAILER_ID,RETAILER,REGION,STATE,CITY
81,W00SARLI,Walmart,South,Arkansas,Little Rock
97,W00SARLI,West Gear,South,Arkansas,Little Rock


In [56]:
df_retailer[
    df_retailer['RETAILER_ID'].isin(duplicate_retailer_ids)
].sort_values('RETAILER_ID')

,RETAILER_ID,RETAILER,REGION,STATE,CITY
63,S00NNENE,Sports Direct,Northeast,New Jersey,Newark
64,S00NNENE,Sports Direct,Northeast,New York,New York
81,W00SARLI,Walmart,South,Arkansas,Little Rock
97,W00SARLI,West Gear,South,Arkansas,Little Rock
84,W00SFLOR,Walmart,Southeast,Florida,Orlando
102,W00SFLOR,West Gear,Southeast,Florida,Orlando
83,W00STEHO,Walmart,South,Texas,Houston
100,W00STEHO,West Gear,South,Texas,Houston


In [57]:
# any RETAILER_ID in df_sales that are not in df_retailer? RESULT: Yes, 1....999999999
df_sales[
    ~df_sales['RETAILER_ID'].isin(df_retailer['RETAILER_ID'])
]['RETAILER_ID'].value_counts()

,count
RETAILER_ID,
999999999,1


In [58]:
# looking at record where RETAILER_ID == 999999999
df_sales[
    df_sales['RETAILER_ID'] == '999999999'
]

,ORDER_ID,RETAILER_ID,INVOICE_DATE,MONTH,DAY,YEAR,PRODUCT_ID,PRICE_PER_UNIT,UNITS_SOLD,OPERATING_MARGIN,SALES_METHOD
1446,8668,999999999,2021-07-23,7,23,2021,20,60.0,298,0.42,Outlet


In [59]:
# updating RETAILER_ID to WOOMWIMI from 999999999
df_sales.loc[
    df_sales['RETAILER_ID'] == '999999999',
    'RETAILER_ID'
] = 'W00MWIMI'

In [60]:
# verify the fix worked
df_sales[
    ~df_sales['RETAILER_ID'].isin(df_retailer['RETAILER_ID'])
]['RETAILER_ID'].value_counts()

,count
RETAILER_ID,


In [61]:
df_retailer[
    df_retailer['RETAILER_ID'].duplicated(keep=False)
].sort_values(['RETAILER_ID', 'RETAILER'])

,RETAILER_ID,RETAILER,REGION,STATE,CITY
63,S00NNENE,Sports Direct,Northeast,New Jersey,Newark
64,S00NNENE,Sports Direct,Northeast,New York,New York
81,W00SARLI,Walmart,South,Arkansas,Little Rock
97,W00SARLI,West Gear,South,Arkansas,Little Rock
84,W00SFLOR,Walmart,Southeast,Florida,Orlando
102,W00SFLOR,West Gear,Southeast,Florida,Orlando
83,W00STEHO,Walmart,South,Texas,Houston
100,W00STEHO,West Gear,South,Texas,Houston


In [62]:
df_retailer[
    df_retailer['RETAILER'].isin(['Walmart', 'West Gear'])
][
    ['RETAILER_ID', 'RETAILER', 'STATE', 'CITY']
].sort_values(['RETAILER', 'RETAILER_ID'])

,RETAILER_ID,RETAILER,STATE,CITY
80,W00NNENE,Walmart,New York,New York
81,W00SARLI,Walmart,Arkansas,Little Rock
84,W00SFLOR,Walmart,Florida,Orlando
82,W00SMIJA,Walmart,Mississippi,Jackson
83,W00STEHO,Walmart,Texas,Houston
85,W00WHAHO,Walmart,Hawaii,Honolulu
86,W00MININ,West Gear,Indiana,Indianapolis
87,W00MIODE,West Gear,Iowa,Des Moines
88,W00MMIST,West Gear,Missouri,St. Louis
89,W00MOHCO,West Gear,Ohio,Columbus


In [63]:
# inspecting results where RETAILER is equal to Sports Direct
df_retailer[
    df_retailer['RETAILER'] == 'Sports Direct'
][
    ['RETAILER_ID', 'RETAILER', 'REGION', 'STATE', 'CITY']
].sort_values('RETAILER_ID')

,RETAILER_ID,RETAILER,REGION,STATE,CITY
54,S00MILCH,Sports Direct,Midwest,Illinois,Chicago
55,S00MMIDE,Sports Direct,Midwest,Michigan,Detroit
56,S00MMOBI,Sports Direct,Midwest,Montana,Billings
57,S00MNEOM,Sports Direct,Midwest,Nebraska,Omaha
58,S00MNOFA,Sports Direct,Midwest,North Dakota,Fargo
59,S00MSOSI,Sports Direct,Midwest,South Dakota,Sioux Falls
60,S00NCOHA,Sports Direct,Northeast,Connecticut,Hartford
61,S00NMABA,Sports Direct,Northeast,Maryland,Baltimore
62,S00NMABO,Sports Direct,Northeast,Massachusetts,Boston
63,S00NNENE,Sports Direct,Northeast,New Jersey,Newark


In [64]:
df_retailer[
    ((df_retailer['STATE'] == 'New Jersey') & (df_retailer['CITY'] == 'Newark')) |
    ((df_retailer['STATE'] == 'New York') & (df_retailer['CITY'] == 'New York'))
].sort_values(['RETAILER', 'STATE'])

,RETAILER_ID,RETAILER,REGION,STATE,CITY
22,F00NNENE,Foot Locker,Northeast,New York,New York
46,K00NNENE,Kohl's,Northeast,New Jersey,Newark
63,S00NNENE,Sports Direct,Northeast,New Jersey,Newark
64,S00NNENE,Sports Direct,Northeast,New York,New York
80,W00NNENE,Walmart,Northeast,New York,New York


In [65]:
df_sales[
    (df_sales['ORDER_ID'] >= 6950) &
    (df_sales['ORDER_ID'] <= 7120)
][
    ['ORDER_ID', 'RETAILER_ID', 'INVOICE_DATE', 'PRODUCT_ID']
].sort_values('ORDER_ID')

,ORDER_ID,RETAILER_ID,INVOICE_DATE,PRODUCT_ID
4376,6950,W00SMIJA,2021-08-06,40
9198,6951,W00SMIJA,2021-08-06,140
1160,6952,W00SMIJA,2021-09-07,20
2770,6953,W00SMIJA,2021-09-07,30
5984,6954,W00SMIJA,2021-09-07,120
...,...,...,...,...
6011,7116,W00SARLI,2021-12-09,120
7618,7117,W00SARLI,2021-12-09,130
4404,7118,W00SARLI,2021-12-09,40
9226,7119,W00SARLI,2021-12-09,140


In [66]:
# copying df_retailer
df_retailer_clean = df_retailer.copy()

In [67]:
ambiguous_ids = ['W00SARLI', 'W00SFLOR', 'W00STEHO']

In [68]:
df_retailer_clean.loc[
    df_retailer_clean['RETAILER_ID'].isin(ambiguous_ids),
    'RETAILER'
] = 'Walmart / West Gear - Ambiguous'

In [69]:
# remove the duplicate rows
df_retailer_clean = df_retailer_clean.drop_duplicates(
    subset='RETAILER_ID',
    keep='first'
)

In [70]:
# inspect the result
df_retailer_clean['RETAILER_ID'].duplicated().sum()

np.int64(0)

####Merging data

In [71]:
# merge products with sales
df_merged = df_sales.merge(
    df_products,
    on='PRODUCT_ID',
    how='left'
)

In [72]:
# verifying result
df_merged.shape

(9648, 12)

In [73]:
# checking that every sales row is matched to a product name
df_merged['PRODUCT_NAME'].isna().sum()

np.int64(0)

In [74]:
# adding in the cleaned retailer table
df_merged = df_merged.merge(
    df_retailer_clean,
    on='RETAILER_ID',
    how='left'
)

In [75]:
# validate result
df_merged.shape

(9648, 16)

In [76]:
# checking that every merged row has a retailer name
df_merged['RETAILER'].isna().sum()

np.int64(0)

####**SUMMARY of INSPECTION and CLEANING**
The three source tables were inspected, cleaned, and merged into a single analysis-ready dataframe.

Key cleaning steps included:
- Replacing two invalid UNITS_SOLD values using the median for comparable records
- Filling two missing PRICE_PER_UNIT values using nearby comparable transactions
- Converting INVOICE_DATE to datetime
- Correcting the SALES_METHOD typo "Ootlet" to "Outlet"
- Correcting one invalid RETAILER_ID
- Addressing duplicated retailer IDs to prevent duplication during the merge
- Preserving ambiguous Walmart/West Gear retailer IDs as an explicit ambiguous category

The final merged dataframe contains 9,648 rows, with no unmatched products or retailers.


In [77]:
# adding a TOTAL_SALES column to df_merged
df_merged['TOTAL_SALES'] = (
    df_merged['PRICE_PER_UNIT'] * df_merged['UNITS_SOLD']
)

In [78]:
# verify the result
df_merged[
    ['PRICE_PER_UNIT', 'UNITS_SOLD', 'TOTAL_SALES']
].head()

,PRICE_PER_UNIT,UNITS_SOLD,TOTAL_SALES
0,50.0,1200,60000.0
1,50.0,1250,62500.0
2,50.0,1220,61000.0
3,50.0,1200,60000.0
4,60.0,1220,73200.0


##**Answering Business Questions**

In [79]:
# total sales by product for 2021
sales_by_product_2021 = (
    df_merged[df_merged['YEAR'] == 2021]
    .groupby('PRODUCT_NAME')['TOTAL_SALES']
    .sum()
    .sort_values(ascending=False)
)

sales_by_product_2021

,TOTAL_SALES
PRODUCT_NAME,
Men's Street Footwear,74667440.0
Women's Apparel,19178278.0
Men's Athletic Footwear,16339593.0
Women's Street Footwear,13535783.0
Men's Apparel,13025293.0
Women's Athletic Footwear,11160629.0


### VP Question 1: Highest-Selling Product Category in 2021

Men's Street Footwear generated the highest sales in 2021, with total sales of $74,667,440.

In [80]:
# VP Question 2
womens_sales_by_state_2021 = (
    df_merged[
        (df_merged['YEAR'] == 2021) &
        (df_merged['PRODUCT_NAME'].str.contains("Women's"))
    ]
    .groupby('STATE')['TOTAL_SALES']
    .sum()
    .sort_values(ascending=False)
)

womens_sales_by_state_2021

,TOTAL_SALES
STATE,
Maine,2176301.0
Delaware,2023575.0
New Hampshire,1916400.0
Arizona,1798900.0
Missouri,1771992.0
Illinois,1743277.0
Virginia,1719886.0
Nebraska,1712680.0
Connecticut,1600156.0


### VP Question 2: Highest State Sales for Women's Products in 2021

Maine generated the highest sales of women's products in 2021, with total sales of $2,176,301.

In [81]:
# VP Question 3
mens_sales_by_state_2021 = (
    df_merged[
        (df_merged['YEAR'] == 2021) &
        (df_merged['PRODUCT_NAME'].str.contains("Men's"))
    ]
    .groupby('STATE')['TOTAL_SALES']
    .sum()
    .sort_values(ascending=False)
)

mens_sales_by_state_2021

,TOTAL_SALES
STATE,
New Hampshire,54208080.0
Delaware,2334300.0
Arizona,2261025.0
Maine,2217190.0
Illinois,2093438.0
Missouri,1951530.0
Connecticut,1926568.0
New York,1847069.0
Nebraska,1718726.0


### VP Question 3: Highest State Sales for Men's Products in 2021

New Hampshire generated the highest sales of men's products in 2021, with total sales of $5,420,808.

In [82]:
# VP Question 4
retailer_units = (
    df_merged[df_merged['YEAR'].isin([2020, 2021])]
    .groupby(['YEAR', 'RETAILER'])['UNITS_SOLD']
    .sum()
    .reset_index()
)

top_retailer_by_year = retailer_units.loc[
    retailer_units.groupby('YEAR')['UNITS_SOLD'].idxmax()
]

top_retailer_by_year

,YEAR,RETAILER,UNITS_SOLD
0,2020,Amazon,317930
5,2021,Foot Locker,1097410


### VP Question 4: Retailer Purchasing the Most Units

- In 2020, Amazon purchased the most units, with 317,930 units.
- In 2021, Foot Locker purchased the most units, with 1,097,410 units.

In [83]:
# overall sales growth 2020 to 2022
sales_by_year = (
    df_merged
    .groupby('YEAR')['TOTAL_SALES']
    .sum()
    .sort_index()
)

sales_by_year

,TOTAL_SALES
YEAR,
2020,24226825.0
2021,147907016.0


In [84]:
# growth rate calculaton
sales_growth_pct = (
    (sales_by_year.loc[2021] - sales_by_year.loc[2020])
    / sales_by_year.loc[2020]
) * 100

sales_growth_pct

np.float64(510.50928464625474)

### Additional Insight 1: Overall Sales Growth

Total sales increased from 24,226,825 in 2020 to 147,907,016 in 2021, an increase of approximately 510.5%.

In [85]:
# sales by method
sales_by_method = (
    df_merged
    .groupby('SALES_METHOD')['TOTAL_SALES']
    .sum()
    .sort_values(ascending=False)
)

sales_by_method

,TOTAL_SALES
SALES_METHOD,
Online,96943319.0
Outlet,39536647.0
In-store,35653875.0


In [86]:
sales_by_method_year = (
    df_merged
    .groupby(['YEAR', 'SALES_METHOD'])['TOTAL_SALES']
    .sum()
    .unstack()
)

sales_by_method_year

SALES_METHOD,In-store,Online,Outlet
YEAR,,,
2020,9379800.0,4519966.0,10327059.0
2021,26274075.0,92423353.0,29209588.0


### Additional Insight 2: Online Sales Growth

Online sales increased from approximately 4.5 million dollars in 2020 to 92.4 million dollars  in 2021. Online became the company's largest sales channel and was the primary driver of overall sales growth between the two years.